src/main/java/games/dominion/actions/Vassal.java

```diff
@@ -0,0 +1,127 @@
+package games.dominion.actions;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import core.AbstractGameState;
+import core.actions.AbstractAction;
+import core.actions.DoNothing;
+import core.components.Deck;
+import core.interfaces.IExtendedSequence;
+import games.dominion.DominionGameState;
+import games.dominion.DominionConstants.DeckType;
+import games.dominion.cards.CardType;
+import games.dominion.cards.DominionCard;
+
+/**
+ * Action to play the Vassal card. This card allows the player to draw the top card of their deck, and if it is an action card, they can choose to play it.
+ */
+public class Vassal extends DominionAction implements IExtendedSequence {
+
+    boolean executed = false;
+    CardType cardType;
+
+    public Vassal(int playerId) {
+        super(CardType.VASSAL, playerId);
+    }
+
+    public Vassal(int playerId, boolean dummy) {
+        super(CardType.VASSAL, playerId, dummy);
+    }
+
+    @Override
+    boolean _execute(DominionGameState state) {
+        Deck<DominionCard> drawPile = state.getDeck(DeckType.DRAW, player);
+        Deck<DominionCard> discardPile = state.getDeck(DeckType.DISCARD, player);
+
+        // Reshuffle if draw pile is empty
+        if (drawPile.getSize() == 0) {
+            discardPile.shuffle(state.getRnd());
+            drawPile.add(discardPile);
+            discardPile.clear();
+        }
+
+        // There is an edge case where this is the only card they have left
+        if (discardPile.getSize() > 0) {
+            state.setActionInProgress(this);
+
+            // Discard top card from the deck...
+            DominionCard card = drawPile.draw();
+
+            // ...and move to table
+            state.getDeck(DeckType.TABLE, player).add(card);
+            cardType = card.cardType();
+        }
+
+        return true;
+    }
+
+    @Override
+    public Vassal copy() {
+        Vassal retValue = new Vassal(player, dummyAction);
+        retValue.executed = executed;
+        retValue.cardType = cardType;
+        return retValue;
+    }
+
+    @Override
+    public List<AbstractAction> _computeAvailableActions(AbstractGameState state) {
+        DominionGameState dgs = (DominionGameState) state;
+
+        DominionCard card = dgs.getDeck(DeckType.TABLE, player).peek();
+
+        assert card.cardType() == cardType: "Discarded CardType: " + card.cardType() + ", does not mach: " + cardType;
```

Comment:
assert statements are ignored when code is running in 'production'. You need to explicitly throw an AssertionErrot here